### PR TITLE
[Persistence] Change root hash calculation to use an SMT

### LINKED ISSUE #834: [Persistence] Change the state root hash computation mechanism
## Objective

The way the root hash is currently computed is as follows:

```go
stateHash := sha256.Sum256(appTreeHash + transactionsTreeHash + ... + flagsTreeHash)
```

This requires the order of the sub tree hashes to be known in order to calculate the state hash, and therefore to prove the inclusion of a specific tree in the state hash calculation this order must be known throughout the codebase.

This should instead be as follows:

1. Create a `rootTree` object in the `TreeStore` module
2. Fill this tree with the following key-value pairings
   - `"app": appTree.Root()`
   - `"transactions": transactionsTree.Root()`
   - ...
3. Calculate the overall state hash by `rootTree.Root()`

This allows for the computation of the state hash in a deterministic manner as the SMT is not order dependent in its updates, thus removing the need to know the tree orders as they are no longer needed. It also allows for the generation of proofs determining whether a state tree hash was included in the root hash - this will be needed for the IBC module's light clients.

## Origin Document

Screenshot 2023-06-16 at 18 49 08

Screenshot 2023-06-16 at 18 49 41

## Goals

- Create a `rootTree` in the `TreeStore`
- Change the logic for state hash computation to insert the different sub trees hashes into the `rootTree` and generate the root this way
- Update the `pokt-network/pocket-network-protocol` documentation to reflect this change

## Deliverable

- [ ] Update state hash calculation to be done via the new `rootTree` SMT
- [ ] Update tests to use the new hashes
- [ ] Update documentation

## Non-goals / Non-deliverables

- Change any other logic in the `TreeStore` module

## General issue deliverables

- [ ] Update the appropriate CHANGELOG(s)
- [ ] Update any relevant local/global README(s)
- [ ] Update relevant source code tree explanations
- [ ] Add or update any relevant or supporting [mermaid](https://mermaid-js.github.io/mermaid/) diagrams

## Testing Methodology

- [ ] **Task specific tests or benchmarks**: `make ...`
- [ ] **New tests or benchmarks**: `make ...`
- [ ] **All tests**: `make test_all`
- [ ] **LocalNet**: verify a `LocalNet` is still functioning correctly by following the instructions at [docs/development/README.md](https://github.com/pokt-network/pocket/tree/main/docs/development)
- [ ] **k8s LocalNet**: verify a `k8s LocalNet` is still functioning correctly by following the instructions [here](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md)

---

**Creator**: @h5law
**Co-Owners**: @dylanlott

### PR DESCRIPTION
**Pull request by h5law**
### Summary generated by Reviewpad on 23 Jun 23 10:35 UTC
This pull request contains several file diffs with various updates and additions. The implementation of the `setupTrees` function in `module.go` has been updated to use a new `stateTree` struct and additional error handling. The `Pocket V1` protocol state hash document has been updated with changes to various infrastructural components and the computation of the state hash. The `persistence_module.go` file has a new `GetTreeStore()` method, and `treestore_module.go` file has a new import statement and `GetTree()` function. Additionally, the `utility/session_test.go` and `state.go` test cases have been updated with new values and logic changes.

## Issue

Fixes #834 

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [x] Documentation
- [ ] Other 

## List of changes

- Update root hash computation to insert different trees hashes and tree name into the tree then compute the root of the tree giving the root hash

## Testing

- [x] `make develop_test`; if any code changes were made
- [x] `make test_e2e` on [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any code changes were made
- [x] `e2e-devnet-test` passes tests on [DevNet](https://pocketnetwork.notion.site/How-to-DevNet-ff1598f27efe44c09f34e2aa0051f0dd); if any code was changed
- [x] [Docker Compose LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md); if any major functionality was changed or introduced
- [x] [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any infrastructure or configuration changes were made

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added, or updated, [`godoc` format comments](https://go.dev/blog/godoc) on touched members (see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))
- [x] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

File: persistence/trees/trees.go [hunk 1]
```diff
@@ -211,7 +249,7 @@ func (t *treeStore) updateMerkleTrees(pgtx pgx.Tx, txi indexer.TxIndexer, height
 			}
 		// Default
 		default:
-			panic(fmt.Sprintf("not handled in state commitment update. Merkle tree #{%v}", treeType))
+			panic(fmt.Sprintf("not handled in state commitment update. Merkle tree: %s", treeName))
```
Comment (Olshansk) [inline persistence/trees/trees.go:252]: Can you replace the panic with a `logger.Fatalf(...)`?

File: persistence/trees/trees_test.go [hunk 2]
```diff
@@ -16,3 +23,28 @@ func TestTreeStore_DebugClearAll(t *testing.T) {
 	// TODO: Write test case for the DebugClearAll method
 	t.Skip("TODO: Write test case for DebugClearAll method")
 }
+
+func TestStateTree_Operations(t *testing.T) {
```
Comment (Olshansk) [inline persistence/trees/trees_test.go:27]: Replace `Operations` with `Basic` or `General` so it's clearer that you just added a single catch all unit test

**Review by Olshansk — CHANGES_REQUESTED** [review]
Couple small nits

File: persistence/trees/trees.go [hunk 3]
```diff
@@ -21,62 +21,61 @@ import (
...
-	appMerkleTree merkleTree = iota
-	valMerkleTree
-	fishMerkleTree
-	servicerMerkleTree
+// stateTree is a wrapper around the SMT that contains an identifying
+// key alongside the tree and nodeStore that backs the tree
+type StateTree struct {
+	key       []byte
+	tree      *smt.SMT
+	nodeStore kvstore.KVStore
```
[the remark this answers is not in the document]
Comment (h5law) [reply persistence/trees/trees.go:65]: name == string(key) will just replace

File: persistence/trees/trees_test.go [hunk 4]
```diff
@@ -23,28 +18,3 @@ func TestTreeStore_DebugClearAll(t *testing.T) {
 	// TODO: Write test case for the DebugClearAll method
 	t.Skip("TODO: Write test case for DebugClearAll method")
 }
-
-func TestStateTree_Operations(t *testing.T) {
```
Comment (Olshansk) [inline persistence/trees/trees_test.go]: Are we bringing this back?

**Comment by Olshansk** [comment]
@dylanlott I think this one is close to approving but figured we should probably merge `persistence/tree-store` to main first and rebase/merge this with main?

Lmk if you have a preference

File: persistence/trees/trees_test.go (hunk 4 above)
Comment (h5law) [reply persistence/trees/trees_test.go]: Not exporting any more those functions dont exist rethinking how we should export trees atm in a way that doesnt alter them

**Review by Olshansk — APPROVED** [review]

**Pull request merged**